ENH Support pipelines in CalibratedClassifierCV

lucyleeow opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Towards #8710 (this is the first/main issue but other issues are mentioned in the comments)


#### What does this implement/fix? Explain your changes.
Don't `_validate_data` if prefit estimator used in `CalibratedClassifierCV`
Includes test

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lucyleeow committed:
wip

lucyleeow committed:
allow prefit

lucyleeow committed:
lint

lucyleeow committed:
remove print

lucyleeow committed:
move check

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +342,24 @@ def decision_function(self, X):
+
+
+def test_calibration_pipeline():
+    """Test that calibration works in pre-fit pipeline with transformer,
```
Use comments instead of docstring. It will not have a nice interaction with pytest when showing the test names

lesteve replied:
Actually this was true with nosetests (that decided that to print the test name when running the test and there was an additional step to figure out the test function) and is fine with pytest but overall not a big deal :shrug: ...

There maybe another thing I am missing of course.

Just for completeness:
```py
# test_docstring.py
def test_bla():
    "docstring for test bla"
    assert 1 + 1 == 2
```

For pytest the name of the test printed on screen is still test_bla
```
❯ pytest test_docstring.py -v
========================================= test session starts ==========================================
platform linux -- Python 3.8.3, pytest-5.4.3, py-1.8.2, pluggy-0.13.1 -- /home/lesteve/miniconda3/envs/del/bin/python
cachedir: .pytest_cache
rootdir: /tmp
collected 1 item                                                                                       

test_docstring.py::test_bla PASSED                                                               [100%]

========================================== 1 passed in 0.00s ===========================================
```

For nosetests the test printed on screen uses the docstring
```
❯ nosetests test_docstring.py -v
docstring for test bla ... ok

----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

glemaitre replied:
OK I am getting old ;)

glemaitre replied:
@lesteve you are right it was my old memories with testing and it was with `nose`

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +342,24 @@ def decision_function(self, X):
+    pipeline.fit(fake_features, labels)
+
+    calib_clf = CalibratedClassifierCV(pipeline, cv='prefit')
+    calib_clf.fit(fake_features, labels)
```
We should probably be checking some fitted attribute from the calibrated classifier to be sure that everything went fine.

glemaitre replied:
For instance, I think that `calib_clf.n_features_in_` will not be defined, isn't it?

We will need to add a property since this is a meta estimator:

```python
    @property
    def n_features_in_(self):
        # For consistency with other estimators we raise a AttributeError so
        # that hasattr() returns False the estimator isn't fitted.
        try:
            check_is_fitted(self)
        except NotFittedError as nfe:
            raise AttributeError(
                "{} object has no n_features_in_ attribute."
                .format(self.__class__.__name__)
            ) from nfe

        return self.calibrated_classifiers_[0].n_features_in_
```

lucyleeow replied:
What does the `from nfe` do here?

connorbrinton replied:
@lucyleeow It sets the `__cause__` of the raised exception to point to the original exception. This allows Python 3 to print the original error message and "The above exception was the direct cause..." in tracebacks, for enhanced debugging. There's more info about it in the [documentation of the `raise` statement](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement) and in [the PEP proposing this feature](https://www.python.org/dev/peps/pep-3134/) 🙂

lucyleeow replied:
Thanks! Good to know!

glemaitre commented on sklearn/calibration.py:
```diff
@@ -176,23 +176,10 @@ def fit(self, X, y, sample_weight=None):
-                                   force_all_finite=False, allow_nd=True)
         X, y = indexable(X, y)
         le = LabelBinarizer().fit(y)
         self.classes_ = le.classes_
```
I am a bit curious about these lines here. In case, that the model has been prefitted, it would be best to use `prefitted_model.classes_` instead? Meaning that we have this part only when the model is not prefitted.

glemaitre commented on sklearn/calibration.py:
```diff
@@ -207,6 +194,22 @@ def fit(self, X, y, sample_weight=None):
+
+            # Check that each cross-validation fold can have at least one
+            # example per class
+            n_folds = self.cv if isinstance(self.cv, int) \
```
can you write an if/else statement without \

```python
if isinstance(self.cv, int):
    n_folds = self.cv
else:
    n_folds = self.cv_n_folds if hasattr(self.cv, "n_folds") else None
```

lucyleeow replied:
`n_folds = self.cv_n_folds if hasattr(self.cv, "n_folds") else None` also doesn't fit on one line so I amended to `if, elif, else` but happy to change.

glemaitre commented on sklearn/calibration.py:
```diff
@@ -207,6 +194,22 @@ def fit(self, X, y, sample_weight=None):
+            # example per class
+            n_folds = self.cv if isinstance(self.cv, int) \
+                else self.cv.n_folds if hasattr(self.cv, "n_folds") else None
+            if n_folds and \
```
Here you can use external `( )` instead of `\`

lucyleeow replied:
Not sure of my formatting here so let me know if not idea.

glemaitre reviewed (commented):
A couple of thoughts to investigate.

cmarmo commented:
Hi @glemaitre and @lucyleeow, there is a PR related to #8710 waiting for review for a year and still alive. It isn't clear to me how those two PRs interact, but may I suggest to have a look at #13060, before moving on with this one? Thanks!

glemaitre commented:
They will be dissociated (even if we are going to have some git conflicts). This one is about internal `sklearn` interoperability while #13060 is about supporting a new type of `y` (i.e. multilabel). I will try to find the time to review #13060 but I am not the most knowledgeable person with multilabel.

glemaitre commented:
So most probably the assert on `n_features_in_` will not be valid with the `DictVectorizer`. I didn't recall the choice we made.

lucyleeow committed:
suggestions

lucyleeow committed:
lint

lucyleeow committed:
lint

lucyleeow committed:
lint

lucyleeow committed:
formatting

lucyleeow committed:
fix attribute

lucyleeow commented:
@glemaitre I'm am not sure about how I addressed this:

> For instance, I think that calib_clf.n_features_in_ will not be defined, isn't it?

* for a NOT prefit estimator, we have to `_validate_data` before we fit and append to the list `self.calibrated_classifiers_`. Thus we can't use `_validate_data(reset=False)`. I've changed back to `reset=True`
* I used your suggested code to set the attribute for prefit estimator in the code `if self.cv == "prefit":`. 

Not sure if this is a good approach so happy to change.

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -202,11 +188,47 @@ def fit(self, X, y, sample_weight=None):
+                n_folds = self.cv.n_folds
+            else:
+                n_folds = None
+            if n_folds and np.any([np.sum(y == class_) < n_folds
+                                  for class_ in self.classes_]):
```
FYI couldn't use temporary var to store `np.any([....` because it doesn't work when `n_folds` is `None`. (but works well in the `and` statement since it wouldn't be evaluated if `n_folds` is `None`)

lucyleeow commented:
ping @glemaitre (the red is codecov)

cmarmo commented on sklearn/calibration.py:
```diff
@@ -202,11 +188,47 @@ def fit(self, X, y, sample_weight=None):
+            # Set `n_features_in_` attribute
+            try:
+                check_is_fitted(self)
+            except NotFittedError as nfe:
+                raise AttributeError(
+                    "{} object has no n_features_in_ attribute."
+                    .format(self.__class__.__name__)
+                ) from nfe
```
I _think_ that this error should be tested.
You might have a look to test coverage for your file [here](https://codecov.io/gh/scikit-learn/scikit-learn/compare/c3835442e50077d32b1b5f9771e7c6b7b999bbfc...e3499ea855599c01cc261077f63d574d4102b152/src/sklearn/calibration.py?before=sklearn/calibration.py). When it's red it means that a test should be added.

lucyleeow replied:
@glemaitre why are we raising an `AttributeError` here? Shouldn't we be raising the `NotFittedError` as this is the cause of `n_features_in_` not appearing..?

glemaitre replied:
It is to simulate that the `n_features_in_` does not exist which would raise an `AttributeError` in a normal estimator.

lucyleeow committed:
wip

lucyleeow committed:
add test

lucyleeow committed:
lint

glemaitre commented on sklearn/calibration.py:
```diff
@@ -202,11 +187,42 @@ def fit(self, X, y, sample_weight=None):
+            else:
+                n_folds = None
+            if n_folds and np.any([np.sum(y == class_) < n_folds
+                                  for class_ in self.classes_]):
```
```suggestion
                                   for class_ in self.classes_]):
```

glemaitre commented on sklearn/calibration.py:
```diff
@@ -202,11 +187,42 @@ def fit(self, X, y, sample_weight=None):
+            check_is_fitted(self.base_estimator)
+            # Set `n_features_in_` attribute
+            try:
+                self.n_features_in_ = self.base_estimator.n_features_in_
```
Since it would happen only for some specific estimator we can try (like `DictVectorizer`):

```python
if hasattr(self, "n_features_in_):
    self.n_features_in_ = self.base_estimator.n_features_in_
```

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +348,31 @@ def decision_function(self, X):
+        ])
+    pipeline.fit(fake_features, labels)
+
+    calib_clf = CalibratedClassifierCV(pipeline, cv='prefit')
```
So if we want to use the `cv=None` with the same pipeline, it is not going to work anyway?

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +348,31 @@ def decision_function(self, X):
+    # Check attributes are obtained from fitted estimator
+    assert_array_equal(calib_clf.classes_, [0, 1])
+    # `DictVectorizer` has no `n_features_in_`
+    msg = "'CalibratedClassifierCV' object has no attribute 'n_features_in_'"
```
I would put this in another test to check `n_features_in_` behaviour. Basically it should give `n_features_in_` of the first estimator of a pipeline or raise an error with weird estimator and should be computed with `cv != "prefit"`

glemaitre reviewed (commented):
We will need an entry in what's new as well

lucyleeow committed:
fix check is fitted, add test

lucyleeow committed:
lint

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +349,53 @@ def decision_function(self, X):
     calibrated_clf.fit(X, y)
+
+
+# Create prefit pipeline
```
If you want to use this several time, it will be cleaner to use a `pytest.fixture`

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +349,53 @@ def decision_function(self, X):
+def test_calibration_pipeline():
+    # Test that calibration works in prefit pipeline with transformer,
+    # where `X` is not array-like, sparse matrix or dataframe at the start.
+    # See issue #8710
```
I would paste directly the github link instead of just the number :)

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +349,53 @@ def decision_function(self, X):
+    calib_clf = CalibratedClassifierCV(pipeline_prefit, cv='prefit')
+    calib_clf.fit(word_data, word_labels)
+    # Check attributes are obtained from fitted estimator
+    assert_array_equal(calib_clf.classes_, [0, 1])
```
```suggestion
    assert_array_equal(calib_clf.classes_, [0, 1])
    assert not hasattr(calib_clf, "n_features_in_")
```

lucyleeow replied:
I test this later in:
```
    calib_clf = CalibratedClassifierCV(pipeline_prefit, cv='prefit')
    msg = "'CalibratedClassifierCV' object has no attribute 'n_features_in_'"
    with pytest.raises(AttributeError, match=msg):
        calib_clf.n_features_in_
```
but I can amend to your way as it is shorter?

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -341,3 +349,53 @@ def decision_function(self, X):
+    assert_array_equal(calib_clf.classes_, [0, 1])
+
+
+def test_calibration_n_features_in_():
```
I think that we can parametrize the test

```suggestion
@pytest.mark.parametrize(
    "clf, cv, fit_model, has_n_features_in",
    [LinearSVC(C=1), 2, False, True]
)
def test_calibration_n_features_in_(clf, cv, fit_model, has_n_features_in):
    ...
    if fit_model:
        clf.fit(X, y)
    ...
        
    assert hasattr(calib_clf, "n_features_in_") is has_n_features_in
```

lucyleeow committed:
fix test, fix pipeline att

lucyleeow commented:
@glemaitre I amended it such that the attribute `n_features_in_` and `classes_` are derived from `base_estimator[-1]` when `base_estimator` is a pipeline.

I think we could/should add a test that checks for when a CV splitter but maybe in another PR since it isn't related to this one?

lucyleeow committed:
add whats new

lucyleeow committed:
merge master

lucyleeow committed:
fix n splits

lucyleeow committed:
add tests

lucyleeow committed:
revert

lucyleeow committed:
fix test

lucyleeow committed:
lint

lucyleeow commented:
I realise why the line:
```
            elif hasattr(self.cv, "n_folds"):
                n_folds = self.cv.n_folds
```
was red. I think it should but `n_splits`. I don't think cv iterators have `n_folds` attribute.

I have fixed it here but let me know if you want me to separate into it's own PR

lucyleeow commented:
I've also added a test for cv iterator and test for default `base_estimator` as codecov was red

lucyleeow committed:
amend test

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -202,11 +188,41 @@ def fit(self, X, y, sample_weight=None):
 
         if self.cv == "prefit":
+            # Set `n_features_in_` attribute
+            if isinstance(self.base_estimator, Pipeline):
+                estimator = self.base_estimator[-1]
+            else:
+                estimator = self.base_estimator
+            check_is_fitted(estimator)
+            self.n_features_in_ = estimator.n_features_in_
+            self.classes_ = estimator.classes_
```
I can't find the discussing about this.

It seems weird that `CalibratedClassiferCV` takes the `n_features_in_` from the final step. If the first step of the Pipeline was a feature selector, then `CalibratedClassiferCV` would not have correct `n_feature_in_`?

Also, for third party estimators in a pipeline, if they do not have `n_feature_in_` or `classes_` this would fail. I would prefer being slightly more lenient:

```py
with suppress(AttributeError):
    self.n_features_in_ = base_estimator.n_features_in_

with suppress(AttributeError):
	self.classes_ = base_estimator.classes_
```

glemaitre replied:
You are right for the `n_features_in_` that it should be the value of the first step and not the last one.

Regarding `classes_`, I would be a bit more conservative. Our `estimator_check` explicitly checks for `classes_` so this is a kind of contract that if you do a scikit-learn classifier, you should have this attribute.

And for `n_features_in_`, we can be lenient in this case because we still don't impose anything yet.

glemaitre replied:
WDYT @thomasjpfan ?

lucyleeow replied:
Yes, you're right. I got mixed up.

lucyleeow replied:
> Regarding classes_, I would be a bit more conservative. Our estimator_check explicitly checks for classes_ so this is a kind of contract that if you do a scikit-learn classifier, you should have this attribute.

As in don't suppress warning when assigning this attribute?

lucyleeow replied:
Also out of interest, which of:
```
if hasattr(calib_clf, "n_features_in_"):
    self.n_features_in_ = base_estimator.n_features_in_
```
and
```
with suppress(AttributeError):
    self.n_features_in_ = base_estimator.n_features_in_
```
would be preferred here?

lucyleeow replied:
I used the first so I don't need to import `suppress` but happy to change

glemaitre replied:
I am actually not sure what is more explicit. I am not used to `suppress` but I find it elegant and it should be more pythonic. Go ahead with it.

thomasjpfan replied:
> WDYT @thomasjpfan ?

I agree with the being more strict with `classes_` and be lenient with `n_features_in_`.

thomasjpfan reviewed (commented):
Thank you for the PR @lucyleeow !

lucyleeow committed:
fix n features in

lucyleeow committed:
use suppres

lucyleeow commented:
thanks @thomasjpfan and @glemaitre, I think I've made the suggested changes

thomasjpfan reviewed (approved):
LGTM Thank you @lucyleeow !

glemaitre reviewed (approved):
LGTM

glemaitre commented:
Thanks @lucyleeow

lucyleeow commented:
Thank you!

odedbd commented:
@lucyleeow I've tried using this (pipeline base estimator, cv='prefit') and now the .fit() call is working, but I ran into a similar issue when I tried to predict, since there is also a data validation step in predict_proba. So I can fit an estimator but I don't see how to use it. What am I missing here?
